developer_documentation/contribution_guidelines.Rmd

```diff
@@ -391,7 +391,40 @@ discuss the following questions:
   prefixes, except that kg is the coherent unit of mass, not g.
 
 * Do not copy and paste code, changing only small parts. Choose a
-  design that eliminates the duplication.
+  design that eliminates the duplication. Duplication is often the
+  result of not separating control flow from data. Consider the
+  following R code, which calculates maximum values of several columns
+  from a data frame called `biocro_result`:
+
+  ```r
+  max_leaf <- max(biocro_result$Leaf)
+  max_stem <- max(biocro_result$Stem)
+  max_root <- max(biocro_result$Root)
+  ```
+
+  The data and behavior can be separated by taking a functional
+  programming approach:
+
+  ```r
+  col_for_max <- c('Leaf', 'Stem', 'Root')
+
+  get_col_max <- function(colname) {
+    max(biocro_result[[colname]])
+  }
+
+  max_vals <- sapply(col_for_max, get_col_max)
+  ```
+
+  In the first example, the behavior is spread throughout all of the
+  code, and there is not an obvious indication of what is being
+  done. Once one understands what is being done, one must still read
+  all of the code to be sure some different behavior is not hidden
+  somewhere.
+
+  In the second example, it is clear that the same task is performed
+  repeatedly because control flow is in a single place, and the
+  place has a meaningful name: `get_col_max`. For a long list, this
+  is succinct, and easier to understand and maintain.
```

Comment:
To my mind, the original code makes it very clear that the same task is being performed repeatedly because we call the function that performs that task, `max`, repeatedly, in three successive lines that have a parallel structure.  True, it _is_ repetitive.  But it isn't unclear.

`get_column_max` or even `get_column_maximum` might be more meaningful names than `get_col_max`, but even these names are a little misleading: the function doesn't take a column, but rather, the name of a column; moreover (in particular), the name of a column of whatever the name `biocro_result` is attached to at the moment when `get_column_max` is called.  (If we want to be really perverse, we can do something like

```r
col_for_max <- c('Leaf', 'Stem', 'Root')

get_col_max <- function(colname) {
    max(biocro_result[[colname]])
}

max_vals <- sapply(col_for_max, get_col_max)

biocro_result <- mtcars

col_for_max <- c('mpg', 'wt', 'hp')

max_vals <- sapply(col_for_max, get_col_max)
```
Probably we shouldn't be encouraging the writing of functions that use global variables.)

Which code is better is highly dependent upon the context.  If we are going to use `max_leaf`, `max_stem`, and `max_root` later on in the code and we don't anticipate needing maximums for any of the other columns, then the original code is probably the best; even though it is a little repetitive, it is probably shorter than any of the alternatives and it saves us from having to do, say,
```r
max_leaf <- max_vals[['Leaf']]
max_stem <- max_vals[['Stem']]
max_root <- max_vals[['Root']]
```
after we are done calling `sapply`.

On the other hand, if all we want to do is display a labeled list of maximums for various columns, the most succinct way of doing that is simply to call
```r
sapply(biocro_result[c('Leaf', 'Stem', 'Root')], max)
```

If we really think we are going to want to repeatedly get maximums for various sets of columns, then maybe it would be worthwhile to define a function like `get_col_max`, but preferably, a version that doesn't use a global variable.  In fact, if we expect to want to do this with various result data frames, it would be worthwhile defining a closure function.

Reply:
I agree that in this case there isn't anything unclear about the code, although it is repetitive. I would have used `max_values = sapply(biocro_results[c('Leaf', 'Stem', 'Root')], max)` or a for loop.

Maybe we can change the example to that? I don't think we need to spend a lot of time on this.